rust/src/ike/parser.rs

```diff
@@ -452,75 +435,70 @@ fn get_group_description(v: u16) -> AttributeValue {
+        let attr = SaAttribute {
+            attribute_format: format.0,
+            attribute_type: get_attribute_type(format.1),
+            attribute_value : match format.1 {
+                1 => get_encryption_algorithm(attribute_length_or_value),
+                2 => get_hash_algorithm(attribute_length_or_value),
+                3 => get_authentication_method(attribute_length_or_value),
+                4 => get_group_description(attribute_length_or_value),
+                11 => match attribute_length_or_value {
+                    1 => AttributeValue::LifeTypeSeconds,
+                    2 => AttributeValue::LifeTypeKilobytes,
+                    _ => AttributeValue::Unknown
+                }
+                _ => AttributeValue::Unknown
+            },
+            numeric_value: match format.0 {
+                1 => Some(attribute_length_or_value as u32),
+                0 => {
+                    numeric_variable_value
+                },
+                _ => None,
+            },
+            hex_value: match format.0 {
+                0 => {
+                    variable_attribute_value.map(|_variable_attribute_value| to_hex(_variable_attribute_value))
+                }
+                _ => None,
+            }
+        };
+        Ok((i, attr))
+    }
+    many0(complete(parse_attribute))(i)
 }
 
 pub fn parse_nonce(i: &[u8], length: u16) -> IResult<&[u8], NoncePayload> {
-    map!(i, take!(length), |v| NoncePayload { nonce_data: v })
-}
-
-named! { pub parse_ikev1_payload_list<&[u8], Vec<IsakmpPayload>>,
-    many0!(
-        complete!(
-            do_parse!(
-                next_payload: be_u8 >>
-                reserved: be_u8 >>
-                payload_length: be_u16 >>
-                payload_data: cond!(payload_length >= 4, take!(payload_length - 4)) >>
-                (
-                    IsakmpPayload {
-                        payload_header: IsakmpPayloadHeader {
-                            next_payload,
-                            reserved,
-                            payload_length
-                        },
-                        data: if let Some(_data) = payload_data { _data } else { b"" }
-                    }
-                )
-            )
-        )
-    )
+    map(take(length), |v| NoncePayload { nonce_data: v })(i)
+}
+
+pub fn parse_ikev1_payload_list(i: &[u8]) -> IResult<&[u8], Vec<IsakmpPayload>> {
+    fn parse_payload(i: &[u8]) -> IResult<&[u8], IsakmpPayload> {
+        let (i, next_payload) = be_u8(i)?;
+        let (i, reserved) = be_u8(i)?;
+        let (i, payload_length) = be_u16(i)?;
+        let (i, payload_data) = cond(payload_length >= 4, |b| take(payload_length - 4)(b))(i)?;
+        Ok((i, IsakmpPayload {
+            payload_header: IsakmpPayloadHeader {
+                next_payload,
+                reserved,
+                payload_length
+            },
+            data: if let Some(_data) = payload_data { _data } else { b"" }
```

Comment:
One could use `unwrap_or_default()` here as well, right?

Reply:
@chifflier do you want to address this or shall we stage for merge as is?

Reply:
This is indeed equivalent to `unwrap_or_default()`. I'd prefer to use the later (shorter and more idiomatic).
Do you want me to push a second commit on this PR, or resubmit?

Reply:
can you do a new PR?

Reply:
Sure! PR sent in #6969